// @kadena/pactjs@0.1.1/lib/PactNumber.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PactNumber = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
/**
 * Constructs a bignumber.js instance and formats into Pact number formats.
 *
 * @alpha
 */
class PactNumber {
    constructor(value) {
        if (isNaN(Number(value)))
            throw new Error('Value is NaN');
        this._number = new bignumber_js_1.default(value);
    }
    toInteger() {
        if (!this._number.isInteger()) {
            throw new Error('PactNumber is not an integer');
        }
        return this._number.toString();
    }
    toStringifiedInteger() {
        return JSON.stringify(this.toInteger());
    }
    toPactInteger() {
        if (!this._number.isInteger()) {
            throw new Error('PactNumber is not an integer');
        }
        return {
            int: this.toInteger(),
        };
    }
    toDecimal() {
        if (this._number.isInteger()) {
            return `${this._number.toString()}.0`;
        }
        return `${this._number.toString()}`;
    }
    toStringifiedDecimal() {
        return JSON.stringify(this.toDecimal());
    }
    toPactDecimal() {
        return {
            decimal: this.toDecimal(),
        };
    }
}
exports.PactNumber = PactNumber;
